File: src/main/java/org/example/trie/Trie.py
from org.example.TrieNode import TrieNode


class Trie(object):

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        cur = self.root

        for c in word:
            if c not in cur.children:
                cur.children[c] = TrieNode()
            cur = cur.children[c]
        cur.endOfWord = True

    def search(self, word):
        cur = self.root

        for c in word:
            if c not in cur.children:
                return False
            cur = cur.children[c]
        return cur.endOfWord

    def startsWith(self, prefix):
        cur = self.root

        for c in prefix:
            if c not in cur.children:
                return False
            cur = cur.children[c]
        return True


# Your Trie object will be instantiated and called as such:
obj = Trie()
word = "apple"
obj.insert(word)
param_2 = obj.search(word)
print(param_2)

prefix = "app"
print(obj.search(prefix))

param_3 = obj.startsWith(prefix)
print(param_3)

File: src/main/java/org/example/trie/Trie.kt
package org.example.trie

class TrieNode{
    val children: MutableMap<Char, TrieNode> = mutableMapOf()
    var endOfWord = false
}

class Trie() {
    val root = TrieNode()

    fun insert(word: String) {
        var cur = root
        for (c in word) {
            if (!cur.children.containsKey(c)) {
                cur.children[c] = TrieNode()
            }
            cur = cur.children[c]!!
        }
        cur.endOfWord = true
    }

    fun search(word: String): Boolean {
        var cur = root
        for (c in word) {
            if (!cur.children.containsKey(c)) return false
            cur = cur.children[c]!!
        }
        return cur.endOfWord
    }

    fun startsWith(prefix: String): Boolean {
        var cur = root
        for (c in prefix) {
            if (!cur.children.containsKey(c)) return false
            cur = cur.children[c]!!
        }
        return true
    }

}